presto-main/src/main/java/com/facebook/presto/operator/repartition/OptimizedPartitionedOutputOperator.java

```diff
@@ -0,0 +1,759 @@
+
+        private final PartitionBuffer[] partitionBuffer;
+        private final List<Type> sourceTypes;
+        private final List<Integer> variableWidthChannels;
+        private final int fixedWidthRowSize;
+        private final DecodedBlockNode[] decodedBlocks;
+
+        private boolean hasAnyRowBeenReplicated;
+
+        public PagePartitioner(
+                PartitionFunction partitionFunction,
+                List<Integer> partitionChannels,
+                List<Optional<ConstantExpression>> partitionConstants,
+                boolean replicatesAnyRow,
+                OptionalInt nullChannel,
+                OutputBuffer outputBuffer,
+                PagesSerdeFactory serdeFactory,
+                List<Type> sourceTypes,
+                DataSize maxMemory,
+                Lifespan lifespan)
+        {
+            this.partitionFunction = requireNonNull(partitionFunction, "pagePartitioner is null");
+            this.partitionChannels = requireNonNull(partitionChannels, "partitionChannels is null");
+            this.partitionConstants = requireNonNull(partitionConstants, "partitionConstants is null").stream()
+                    .map(constant -> constant.map(ConstantExpression::getValueBlock))
+                    .collect(toImmutableList());
+            this.replicatesAnyRow = replicatesAnyRow;
+            this.nullChannel = requireNonNull(nullChannel, "nullChannel is null");
+            this.outputBuffer = requireNonNull(outputBuffer, "outputBuffer is null");
+            this.serde = requireNonNull(serdeFactory, "serdeFactory is null").createPagesSerde();
+
+            int partitionCount = partitionFunction.getPartitionCount();
+
+            int pageSize = max(1, min(DEFAULT_MAX_PAGE_SIZE_IN_BYTES, toIntExact(maxMemory.toBytes()) / partitionCount));
+
+            partitionBuffer = new PartitionBuffer[partitionCount];
+            for (int i = 0; i < partitionCount; i++) {
+                partitionBuffer[i] = new PartitionBuffer(i, sourceTypes.size(), pageSize, pagesAdded, rowsAdded, serde, lifespan);
+            }
+
+            this.sourceTypes = sourceTypes;
+            decodedBlocks = new DecodedBlockNode[sourceTypes.size()];
+
+            ImmutableList.Builder<Integer> variableWidthChannels = ImmutableList.builder();
+            int fixedWidthRowSize = 0;
+            for (int i = 0; i < sourceTypes.size(); i++) {
+                int bytesPerPosition = getFixedWidthTypeSize(i);
+                fixedWidthRowSize += bytesPerPosition;
+
+                if (bytesPerPosition == 0) {
+                    variableWidthChannels.add(i);
+                }
+            }
+            this.variableWidthChannels = variableWidthChannels.build();
+            this.fixedWidthRowSize = fixedWidthRowSize;
+        }
+
+        public ListenableFuture<?> isFull()
+        {
+            return outputBuffer.isFull();
+        }
+
+        public long getInitialRetainedSizeInBytes()
```

Comment:
@arhimondr At the time getInitialRetainedSizeInBytes() is called, the decodedBlocks were not populated yet. Therefore if we call getRetainedSizeInBytes() when OptimizedPartitionedOutputOperator is just created, we will have to change getRetainedSizeInBytes() to the following:

```
        public long getRetainedSizeInBytes()
        {
            long size = 0;

            for (int i = 0; i < decodedBlocks.length; i++) {
                size += decodedBlocks[i] == null? 0 : decodedBlocks[i].getRetainedSizeInBytes();
            }

            for (int i = 0; i < partitionBuffer.length; i++) {
                size += partitionBuffer[i].getRetainedSizeInBytes();
            }

            return size;
        }
```
This adds an additional check on each decodedBlock. Since getRetainedSizeInBytes() is already expensive we want to make it as fast as possible, so I extracted getInitialRetainedSizeInBytes() as a separate method.

Reply:
I ran the benchmark and the difference was little. So I made the change you suggested.

Reply:
Yeah, usually Java is good at the explicit null checks elision. The null check is done by a signal handler.